### DISCUSSION TITLE
Hey, had a few questions around the publishing a stream?

### DISCUSSION THREAD

**Discussion by ajay-collab** [bluenviron/mediamtx#257] [Questions]
You use ffmpeg as your publisher, I've been looking to find a way to publish images as my streams. Basically I would want to have a process where I stream video via a file or network and capture those frames and do some processing and publish them on the server. Could you please let me know how to achieve this?

**Comment by aler9** [comment]
Hi, anything published on the server must be encoded in a streaming format; video and audio formats (h264, ogg, mp3, aac) are streaming formats, while image formats are not, therefore if you want to publish images you have to convert them to video.
In my opinion there are two ways to achieve this:
* output directly video from your script, which is easier. There are a lot of libraries to build videos from image buffers, and send them to a RTSP server, the most common being OpenCV, which have bindings for C++ and Python
* convert images to video with an external program, like ffmpeg. It's difficult because you have to produce images with a constant rate. A simple google search is enough to find how to perform this task (https://stackoverflow.com/questions/24961127/how-to-create-a-video-from-images-with-ffmpeg)

Hope this helps

**Comment by ajay-collab** [comment]
Hmm, the problem is I would have opencv read frames from rtsp uri and then I'm not sure how to encode them so that I could push it to this. I right now have the only option to push images not videos. Any idea around this?

**Comment by aler9** [comment]
Hello, now there's a new option to push frames into rtsp-simple-server, that consists in using gstreamer in this way:
opencv --> rtspclientsink element of gstreamer --> rtsp-simple-server

You can start from an opencv-to-gstreamer example, like this:
https://stackoverflow.com/questions/37339184/how-to-write-opencv-mat-to-gstreamer-pipeline

Then replace
```
mpegtsmux ! udpsink host=localhost port=9999
```
with
```
rtspclientsink location=rtsp://localhost:8554/mystream
```
and after some tries it should work.

**Comment by Kiwa21** [comment]
Hello @aler9 , thanks for the explanation !
 
So you are saying : opencv --> rtspclientsink element of gstreamer --> rtsp-simple-server

For example, a python opencv code would look something like this : 

`gst_str_rtp = "appsrc ! videoconvert ! x264enc noise-reduction=10000 tune=zerolatency byte-stream=true threads=4 ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream"
 rtsp_writer = cv2.VideoWriter(gst_str_rtp, 0, fps, (w, h), True)
 rtsp_writer.write(im0)`

Then rtsp-simple-server.yml : 

```
rtspAddress: :8554

paths:
  mystream:
    source: publisher
    sourceProtocol: automatic
```

And that's it ?
I tried and I have nothing published so probably an issue somewhere.

Thanks !

**Reply by aler9** [reply]
@Kiwa21 OpenCV must be compiled with GStreamer support, take a look at the readme here: https://github.com/aler9/rtsp-simple-server#opencv

**Reply by florianguyot** [reply]
@Kiwa21 did you manage to make run ?

**Reply by mheriyanto** [reply]
Hello @Kiwa21, I have faced same problem with you. Any updates for your issue? Thanks.

**Reply by Kiwa21** [reply]
Hello, I don't remember exactly but I think I made it work with
`
fps, w, h = 30, im0.shape[1], im0.shape[0]

gst_str_rtp = "appsrc ! videoconvert ! x264enc noise-reduction=10000 tune=zerolatency byte-stream=true threads=4 ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream"

rtsp_writer = cv2.VideoWriter(gst_str_rtp, 0, fps, (w, h), True)

rtsp_writer.write(im0)
`

Be sure to have the mystream path defined in rtsp-simple-server config.


In my case, I decided to use the VidGear libraries (which are very similar to what I wrote above)

output_params = {"-f":"rtsp","-rtsp_transport":"tcp","-vcodec":"libx264","-profile:v":"main","-preset:v":"fast","-g":60,"-keyint_min":60,"-sc_threshold":0,"-b:v":"2500k","-maxrate":"2500k","-bufsize":"2500k", "-f":"flv"}              
vidgear_writer = WriteGear(output_filename = 'rtsp://localhost:8554/mystream', logging =True, **output_params)
vidgear_writer.write(im0)

**Reply by mheriyanto** [reply]
Hi @Kiwa21 , Thank you very much for your guidance, I appreciate it. First one, I have tried using 

`gst_str_rtp = "appsrc ! videoconvert ! x264enc noise-reduction=10000 tune=zerolatency byte-stream=true threads=4 ! h264parse ! rtspclientsink location=rtsp://localhost:8554/mystream"` 

but how to define path in rtsp-simple-server config?
I follow your rtsp-simple-server.yml above but I can't open this url `rtsp://localhost:8554/mystream` on VLC media player.

Second one, using VidGear library as your guidance I got error below

```
Input #0, rawvideo, from 'pipe:':
  Duration: N/A, start: 0.000000, bitrate: 552960 kb/s
    Stream #0:0: Video: rawvideo (BGR[24] / 0x18524742), bgr24, 1280x720, 552960 kb/s, 25 tbr, 25 tbn, 25 tbc
rtsp://localhost:8554/mystream: Protocol not found
10:19:05 ::   WriteGear   ::  ERROR   :: BrokenPipeError caught, Wrong values passed to FFmpeg Pipe, Kindly Refer Docs!
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/dist-packages/vidgear/gears/writegear.py", line 354, in write
    self.__process.stdin.write(frame.tostring())
BrokenPipeError: [Errno 32] Broken pipe
```
I tried some options for this issue but it's still error.

**Reply by aler9** [reply]
OpenCV instructions in the README have been updated in order to further help users.

**Reply by mheriyanto** [reply]
Hi @aler9, I followed your tutorial [here](https://github.com/aler9/rtsp-simple-server#from-opencv) and used default rtsp-simple-server.yml from [here](https://github.com/aler9/rtsp-simple-server/releases/download/v0.18.0/rtsp-simple-server_v0.18.0_linux_amd64.tar.gz), it works but I got blank video, there is no red rectangle on frame when I opened `rtsp://localhost:8554/mystream` on VLC media player.

[a comment by an automated account is not shown]
